### DISCUSSION TITLE
Is there a limit to packages built in the CI?

### DISCUSSION THREAD

**Discussion by caseybaggz** [codesandbox/codesandbox-client#6807] [Q&A]
Hello! I have a mono-repo using Yarn Berry Workspaces with Pnp and we have successfully been using the ci.json config for the past year. However, our repo now has 4 workspaces we need to build but our config only continues to build the original three.

### Question
Is there a limit to how many packages the `ci.json` file will build (like three)? 

I can't seem to find any information on that but the ci docs are kind of minimal in general.

**Comment by lbogdan** [comment]
Can you please share the repository?

**Reply by caseybaggz** [reply]
Absolutely: 
https://github.com/pluralsight/tva

**Reply by caseybaggz** [reply]
@lbogdan You can check out the links in the current PRs to see that the CS builds only have our 3 workspaces and not 4 in the deps.

**Reply by lbogdan** [reply]
How many packages should we build? I see 4 in the [latest PR build](https://ci.codesandbox.io/status/pluralsight/tva/pr/471/builds/275992):

![image](https://user-images.githubusercontent.com/2083930/181773599-57fb9151-e10c-4341-ac46-35d37c1c43c2.png)

**Reply by caseybaggz** [reply]
Huh, that's correct but if you go to the actual sandbox it always leaves out `react-utils`?

**Reply by caseybaggz** [reply]
Wait...I swear it was leaving it off. Now it's working.
https://codesandbox.io/s/ps-react-0oyf0q

Well, apparently I have wasted your time. I'm so so so so sorry! 🙈

**Reply by caseybaggz** [reply]
Oh...it's only using all four on the default React sandbox and for all of our custom ones it's leaving it off. OK, I knew there was something.

https://codesandbox.io/s/ps-react-typescript-yrrhdt

**Reply by lbogdan** [reply]
No worries, glad that it works! 🙂

**Reply by caseybaggz** [reply]
Ah...I *think* I found the possible issue (don't know if it's a bug or intended but should probably mention it in the CI docs).

The React custom sandbox had no other workspace npm packages in the deps list. All the other custom sandboxes did. So it's as if the ci.json job just completely ignored replacing the workspace deps in general if one existed.

**Working Example:**
1. Run ci.json (CI)
2. CI starts building all workspaces
3. CI uses sandbox defined
4. CI adds all workspace packages in the custom sandbox as npm deps
5. It works

**Bug Example (pre-defined workspaces):**
1. Run ci.json (CI)
2. CI starts building all workspaces
3. CI uses sandbox defined
4. CI adds all workspace packages in the custom sandbox as npm deps
5. CI notices npm packages already exist in custom sandbox and ingores step & exit

☝️ I feel as though step 5 of the Bug example should instead do this...
5. CI notices npm packages already exist in custom sandbox and replaces version with commit
6. CI adds any remaining missing workspace packages as npm packages to deps list
7. it works

**Reply by lbogdan** [reply]
This is indeed expected behavior. If some of the built packages already exist in `dependencies`, we only replace those, and also all their dependencies on other built packages. As an example:
- repository builds packages `foo` and `bar`, `foo` depends on `bar`
- sandbox only has `foo` in `dependencies`
- we only replace `foo` with `foo-build` in the sandbox, and also `bar` with `bar-build` in `foo-build`'s `dependencies`

If you want to also have `bar` in the resulting sandbox's `dependencies`, you should add it to the source sandbox's `dependencies`.

**Reply by caseybaggz** [reply]
Ahhhh, yeah, that should definitely be on the docs ci page somewhere. If the website is an open-source repository, I don't mind updating that. Just let me know. 👍

**Reply by lbogdan** [reply]
We recently moved them to a separate, (for now) private repository, but if you want to make a PR for [the initial version](https://github.com/codesandbox/codesandbox-client/blob/master/packages/homepage/content/docs/8-ci.md), which, for CI, we just copied over, we can probably apply it.

**Comment by aj-foster** [comment]
Psst... Hi Casey 👋🏼

**Reply by caseybaggz** [reply]
@aj-foster Hello! 👋
